wrangler-service/src/main/java/io/cdap/wrangler/service/directive/WorkspaceUpgrader.java

```diff
@@ -0,0 +1,140 @@
+
+import io.cdap.cdap.api.NamespaceSummary;
+import io.cdap.cdap.api.service.http.SystemHttpServiceContext;
+import io.cdap.cdap.etl.proto.ArtifactSelectorConfig;
+import io.cdap.cdap.etl.proto.connection.ConnectorDetail;
+import io.cdap.cdap.etl.proto.connection.SpecGenerationRequest;
+import io.cdap.cdap.spi.data.transaction.TransactionRunners;
+import io.cdap.wrangler.PropertyIds;
+import io.cdap.wrangler.api.Row;
+import io.cdap.wrangler.dataset.workspace.Workspace;
+import io.cdap.wrangler.dataset.workspace.WorkspaceDataset;
+import io.cdap.wrangler.proto.NotFoundException;
+import io.cdap.wrangler.proto.workspace.v2.Artifact;
+import io.cdap.wrangler.proto.workspace.v2.Plugin;
+import io.cdap.wrangler.proto.workspace.v2.SampleSpec;
+import io.cdap.wrangler.proto.workspace.v2.StageSpec;
+import io.cdap.wrangler.proto.workspace.v2.WorkspaceDetail;
+import io.cdap.wrangler.proto.workspace.v2.WorkspaceId;
+import io.cdap.wrangler.store.upgrade.UpgradeStore;
+import io.cdap.wrangler.store.workspace.WorkspaceStore;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * Upgrader for workspaces. This upgrade should be run after the {@link ConnectionUpgrader}
+ */
+public class WorkspaceUpgrader {
+  private static final Logger LOG = LoggerFactory.getLogger(WorkspaceUpgrader.class);
+
+  private final UpgradeStore upgradeStore;
+  private final SystemHttpServiceContext context;
+  private final long upgradeBeforeTsSecs;
+  private final ConnectionDiscoverer discoverer;
+  private final WorkspaceStore wsStore;
+
+  public WorkspaceUpgrader(UpgradeStore upgradeStore, SystemHttpServiceContext context, long upgradeBeforeTsSecs,
+                           WorkspaceStore wsStore) {
+    this.upgradeStore = upgradeStore;
+    this.context = context;
+    this.upgradeBeforeTsSecs = upgradeBeforeTsSecs;
+    this.discoverer = new ConnectionDiscoverer(context);
+    this.wsStore = wsStore;
+  }
+
+  public void upgradeWorkspaces() throws Exception {
+    List<NamespaceSummary> namespaces = context.listNamespaces();
+    for (NamespaceSummary ns : namespaces) {
+      if (!upgradeStore.isWorkspaceUpgradeComplete(ns)) {
+        upgradeWorkspacesInConnections(ns);
+      }
+    }
+    upgradeStore.setWorkspaceUpgradeComplete();
+  }
+
+  private void upgradeWorkspacesInConnections(NamespaceSummary namespace) {
+    List<Workspace> workspaces = TransactionRunners.run(context, ctx -> {
+      WorkspaceDataset wsDataset = WorkspaceDataset.get(ctx);
+      return wsDataset.listWorkspaces(namespace, upgradeBeforeTsSecs);
+    });
+
+    for (Workspace workspace : workspaces) {
+      List<Row> sample = Collections.emptyList();
+      try {
+        sample = DirectivesHandler.fromWorkspace(workspace);
+      } catch (Exception e) {
+        // this should not happen, but guard here to avoid failing the entire upgrade process
```

Comment:
ok looked at the fromWorkspace code, it throws exceptions if it can't deserialize, which shouldn't happen.

In either case, it's better for the log warning to be more specfic, like

"Could not decode the sample data for workspace {}. This workspace will not be upgraded"